### PR TITLE
NN-3117 use complexity endpoint

### PR DESCRIPTION
**Pull request by JamieDraperUK**
It's missing integration tests at the minute but will add some in future PRs

### CODE REVIEW

File: src/test/java/uk/gov/justice/digital/hmpps/keyworker/services/ComplexityOfNeedServiceTest.kt [hunk 1]
```diff
@@ -16,43 +19,86 @@ import uk.gov.justice.digital.hmpps.keyworker.events.ComplexityOfNeedListenerTes
...
   fun `should deallocate offenders that have high complexity of needs`() {
-    complexityOfNeedService.onComplexityChange(OFFENDER_NO, ComplexityOfNeedLevel.HIGH)
+    complexityOfNeedService.onComplexityChange(OFFENDER_NO_1, ComplexityOfNeedLevel.HIGH)
+
+    verify(keyworkerService, times(1)).deallocate(OFFENDER_NO_1)
+  }
+
+  @Test
+  fun `should not filter out complex offenders for none enabled prisons`() {
+    val noneComplexOffenders = complexityOfNeedService.getComplexOffenders("LEI", setOf(OFFENDER_NO_1))
```
Comment (sp-watson) [inline src/test/java/uk/gov/justice/digital/hmpps/keyworker/services/ComplexityOfNeedServiceTest.kt:64]: A bit confusing to call these `non` complex when the method call is complex?!

**Review by JamieDraperUK — COMMENTED** [review]
Agreed!

File: src/test/java/uk/gov/justice/digital/hmpps/keyworker/services/ComplexityOfNeedServiceTest.kt [hunk 2]
```diff
@@ -16,43 +19,86 @@ import uk.gov.justice.digital.hmpps.keyworker.events.ComplexityOfNeedListenerTes
...
+
+    verify(complexityOfNeedAPI, times(1)).getOffendersWithMeasuredComplexityOfNeed(setOf(OFFENDER_NO_1))
+  }
+
+  @Test
+  fun `should remove all offenders that have a high complexity of need`() {
+    whenever(complexityOfNeedAPI.getOffendersWithMeasuredComplexityOfNeed(any())).thenReturn(
+      listOf(
+        ComplexOffender(OFFENDER_NO_1, ComplexityOfNeedLevel.HIGH),
+        ComplexOffender(OFFENDER_NO_2, ComplexityOfNeedLevel.LOW)
```
Comment (sp-watson) [inline src/test/java/uk/gov/justice/digital/hmpps/keyworker/services/ComplexityOfNeedServiceTest.kt:86]: Did you mean to include OFFENDER_NO_2 in this test? Just checking!

File: src/test/java/uk/gov/justice/digital/hmpps/keyworker/services/ComplexityOfNeedServiceTest.kt (hunk 1 above)
Comment (sp-watson) [inline src/test/java/uk/gov/justice/digital/hmpps/keyworker/services/ComplexityOfNeedServiceTest.kt]: This is confusing... maybe better to name the variable "noneEnabledPrisonOffenders" or something.

Actually this indicates that there won't be any auto-allocation of offenders in prisons that aren't enabled - is that right?

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/health/ComplexityOfNeedApiHealth.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression
+import org.springframework.http.ResponseEntity
+import org.springframework.stereotype.Component
+import org.springframework.web.reactive.function.client.WebClient
+import org.springframework.web.reactive.function.client.WebClientResponseException
+import java.time.Duration
+
+@Component
+@ConditionalOnExpression("T(org.apache.commons.lang3.StringUtils).isNotBlank('\${complexity_of_need_uri}')")
+class ComplexityOfNeedApiHealth(
```
Comment (sp-watson) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/health/ComplexityOfNeedApiHealth.kt:15]: is it worth writing a test for this?

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/ComplexityOfNeedService.kt [hunk 4]
```diff
@@ -2,21 +2,40 @@ package uk.gov.justice.digital.hmpps.keyworker.services
...
+
+    val complexOffenders = complexityOfNeedAPI
+      .getOffendersWithMeasuredComplexityOfNeed(offenders)
+      .filter {
+        it.level.equals(ComplexityOfNeedLevel.HIGH)
+      }.map {
+        it.offenderNo
+      }.toSet()
+
+    return offenders.filter {
```
Comment (sp-watson) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/ComplexityOfNeedService.kt:31]: I think you can use offenders.retainAll(complexOffenders) here

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/ComplexityOfNeedService.kt [hunk 5]
```diff
@@ -2,21 +2,40 @@ package uk.gov.justice.digital.hmpps.keyworker.services
...
   val telemetryClient: TelemetryClient
 ) {
   companion object {
     private val log = LoggerFactory.getLogger(this::class.java)
   }
 
+  fun getComplexOffenders(prisonId: String, offenders: Set<String>): Set<String> {
+    if (!enabledPrisons.contains(prisonId)) return offenders
+
+    val complexOffenders = complexityOfNeedAPI
```
Comment (sp-watson) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/ComplexityOfNeedService.kt:23]: Maybe rename to highComplexityOffenders or even pass in the complexity threshold to this function?

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/ComplexityOfNeedAPI.kt [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  val offenderNo: String,
+  val level: ComplexityOfNeedLevel,
+  val sourceUser: String? = null,
+  val sourceSystem: String? = null,
+  val notes: String? = null,
+  val createdTimeStamp: LocalDateTime? = null
+)
+
+@Service
+class ComplexityOfNeedAPI(private val complexityOfNeedWebClient: WebClient) {
```
Comment (sp-watson) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/ComplexityOfNeedAPI.kt:20]: Don't know if the "*API" suffix is standard naming, but at previous places I have seen these server classes called "*Gateway" or "*Client"

**Review by sp-watson — APPROVED** [review]

**Pull request merged**